test/model/autocomplete_test.dart

```diff
@@ -349,4 +320,131 @@ void main() {
+          RecentDmConversation(userIds: [1], maxMessageId: 100),
+        ]);
+
+        final userA = eg.user(userId: 1);
+        final userB = eg.user(userId: 2);
+        final compareAB = MentionAutocompleteView.compareByDms(userA, userB, store: store);
+        check(compareAB).isNegative();
+      });
+
+      test('has DMs with userB but not userA, prioritizes userB', () async {
+        await prepare(dmConversations: [
+          RecentDmConversation(userIds: [2], maxMessageId: 100),
+        ]);
+
+        final userA = eg.user(userId: 1);
+        final userB = eg.user(userId: 2);
+        final compareAB = MentionAutocompleteView.compareByDms(userA, userB, store: store);
+        check(compareAB).isGreaterThan(0);
+      });
+
+      test('doesn\'t have DMs with userA or userB, prioritizes neither', () async {
+        await prepare(dmConversations: []);
+
+        final userA = eg.user(userId: 1);
+        final userB = eg.user(userId: 2);
+        final compareAB = MentionAutocompleteView.compareByDms(userA, userB, store: store);
+        check(compareAB).equals(0);
+      });
+    });
+
+    test('autocomplete suggests relevant users in the following order: '
+      '1. Users most recent in the DM conversations', () async {
+      final users = [
+        eg.user(userId: 0),
+        eg.user(userId: 1),
+        eg.user(userId: 2),
+        eg.user(userId: 3),
+        eg.user(userId: 4),
+      ];
+
+      final dmConversations = [
+        RecentDmConversation(userIds: [3],    maxMessageId: 300),
+        RecentDmConversation(userIds: [0],    maxMessageId: 200),
+        RecentDmConversation(userIds: [0, 1], maxMessageId: 100),
+      ];
+
+      Future<void> checkResultsIn(Narrow narrow, {required List<int> expected}) async {
+        await prepare(users: users, dmConversations: dmConversations);
+        final view = MentionAutocompleteView.init(store: store, narrow: narrow);
+
+        bool done = false;
+        view.addListener(() { done = true; });
+        view.query = MentionAutocompleteQuery('');
+        await Future(() {});
+        check(done).isTrue();
+        final results = view.results
+          .map((e) => (e as UserMentionAutocompleteResult).userId)
+          .toList();
+        check(results).deepEquals(expected);
+      }
+
+      const streamNarrow = StreamNarrow(1);
+      await checkResultsIn(streamNarrow, expected: [3, 0, 1, 2, 4]);
+
+      const topicNarrow = TopicNarrow(1, 'topic');
+      await checkResultsIn(topicNarrow, expected: [3, 0, 1, 2, 4]);
+
+      final dmNarrow = DmNarrow(allRecipientIds: [eg.selfUser.userId], selfUserId: eg.selfUser.userId);
+      await checkResultsIn(dmNarrow, expected: [3, 0, 1, 2, 4]);
+
+      const allMessagesNarrow = CombinedFeedNarrow();
+      // Results are in the original order as we do not sort them for
+      // [CombinedFeedNarrow] because we can not access autocomplete for now.
+      await checkResultsIn(allMessagesNarrow, expected: [0, 1, 2, 3, 4]);
```

Comment:
These should be separate test cases, because each of them is logically independent of the others — checking something that doesn't need the others to have happened first. Having separate test cases is helpful because it means that if a change breaks some of them, you can immediately see which ones it breaks and doesn't break, rather than have to fiddle around editing the test in order to try each one in turn.

You can keep each of them pretty much just as compact as this by… actually it looks like you've basically already done all the work needed for that, with the `checkResultsIn` helper :slightly_smiling_face:. Just move the `users` and `dmConversations` locals inside that helper, so that they stay inside (transitively) `test` calls rather than being at the top level of a `group`.